services/net/filemonitor/FilemonitorAction.cs

```diff
@@ -0,0 +1,678 @@
+    {
+        reference.Offset = result.Offset;
+        reference.Partition = result.Partition;
+        reference.WorkflowStatus = (int)WorkflowStatus.Received;
+        return await this.Api.UpdateContentReferenceAsync(reference);
+    }
+
+    /// <summary>
+    /// Send AJAX request to api to update content reference.
+    /// </summary>
+    /// <param name="reference"></param>
+    /// <param name="item"></param>
+    /// <returns></returns>
+    private async Task<ContentReferenceModel?> UpdateContentReferenceAsync(ContentReferenceModel reference, SourceContent item)
+    {
+        reference.PublishedOn = item.PublishedOn != DateTime.MinValue ? item.PublishedOn : null;
+        //reference.SourceUpdateOn = item.LastUpdatedTime != DateTime.MinValue ? item.LastUpdatedTime.UtcDateTime : null;
+        return await this.Api.UpdateContentReferenceAsync(reference);
+    }
+
+    /// <summary>
+    /// Get a news item value (e.g. headline, published date, author) from an XML story. Different publications store these values
+    /// in different tags/attributes in the XML document. The keys for the list of supported values is in the Fields class.
+    ///
+    /// Each of these keys maps to an entry in the data source's Connection string which contains the tag/attribute name. If the
+    /// value is stored in a tag, the Connection string entry contains the name of the tag, e.g., for the GLOBE data source
+    /// the story content is in the body.content tag. For the BCNG data source it is in the story tag, so the connection
+    /// entries for those two items are:
+    ///
+    ///     "story": "body.content"
+    ///     "story": "story"
+    ///
+    /// respectively. If the value is stored in a tag's attribute, the Connection string value will have the following format:
+    ///
+    ///     tag-name!attribute-name
+    ///
+    /// For the GLOBE data source the publication date is stored in the date.publication attribute of the pubdata tag for each
+    /// story. This results in the following entry for "date" in the connection string for GLOBE:
+    ///
+    ///     "date": "pubdata!date.publication"
+    ///
+    /// </summary>
+    /// <param name="story"></param>
+    /// <param name="key"></param>
+    /// <param name="dataSource"></param>
+    /// <returns></returns>
+    private string GetXmlData(XmlElement story, string key, DataSourceModel dataSource)
+    {
+        var value = GetConnectionValue(dataSource, key);
+        var result = "";
+
+        if (!String.IsNullOrEmpty(value))
+        {
+            var comps = value.Split('!');
+            if (comps.Count() == 2) // Tag and attribute name present
+            {
+                XmlNodeList nodeList = story.GetElementsByTagName(comps[0]);
+
+                if (nodeList is not null && nodeList.Count > 0 && nodeList[0]!.Attributes is not null && nodeList![0]!.Attributes!.Count > 0)
```

Comment:
This is the first time I've worked with warnings in C#. The documentation indicated that with a statement like:

if (nodeList is not null && nodeList.Count > 0 && nodeList[0]!.Attributes is not null && nodeList![0]!.Attributes!.Count > 0)

The compiler will know you've tested for nulls and will disable the warnings inside the body of the if statement. It didn't, so I disabled them manually. Obviously this is a first attempt, so I'd love some guidance.

Reply:
general rule of thumb if you using `!` to ignore the warning it's probably not really solving the issue.  The compiler is telling you it can't be sure it won't have a null value.  While some times it is wrong and doesn't know, most of the time it is right.

Reply:
Any tips on why the compiler doesn't know I've tested for null?

Reply:
Fixed.